controllers/idler/owners.go

```diff
@@ -0,0 +1,106 @@
+package idler
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ownerFetcher struct {
+	resourceLists   []*metav1.APIResourceList // All available API in the cluster
+	discoveryClient discovery.ServerResourcesInterface
+	dynamicClient   dynamic.Interface
+}
+
+func newOwnerFetcher(discoveryClient discovery.ServerResourcesInterface, dynamicClient dynamic.Interface) *ownerFetcher {
+	return &ownerFetcher{
+		discoveryClient: discoveryClient,
+		dynamicClient:   dynamicClient,
+	}
+}
+
+// getOwners returns the whole tree of all controller owners going recursively to the top owner for the given object
+func (o *ownerFetcher) getOwners(ctx context.Context, obj metav1.Object) ([]client.Object, error) {
+	if o.resourceLists == nil {
+		// Get all API resources from the cluster using the discovery client. We need it for constructing GVRs for unstructured objects.
+		// Do it here once, so we do not have to list it multiple times before listing/getting every unstructured resource.
+		resourceLists, err := o.discoveryClient.ServerPreferredResources()
```

Comment:
So you will discover APIs for every owner fetcher (== every pod in every user namespace per idler run) instead of the current (once per idler run)? I guess it's not a big deal but was curious if you planning to optimize it in the future PRs or leave it as is.

Reply:
The resourceList is retrieved only if it's not already present (and stored) in the  "owners fetcher" object (there is also a unit-test for it). My plan is to create an instance of the owner fetcher per idler run (once in a reconcile), so there will be only one API discovery per idler run - in the same way as it is done now.